java/src/main/java/vw/learner/VWFactory.java

```diff
@@ -0,0 +1,124 @@
+package vw.learner;
+
+import vw.jni.NativeUtils;
+
+import java.io.IOException;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * This is the only entrance point to create a VWLearner.  It is the responsibility of the user to supply the type they want
+ * given the VW command.  If that type is incorrect a {@link java.lang.ClassCastException} is thrown.
+ * @author jmorra
+ */
+final public class VWFactory {
+    private volatile static boolean loadedNativeLibrary = false;
+    private static final Lock STATIC_LOCK = new ReentrantLock();
+
+    enum VWReturnType {
+        Unknown, VWFloatType, VWIntType, VWIntArrayType, VWFloatArrayType
+    }
+
+    private VWFactory() {}
+
+    /**
+     * This is the only way to construct a VW Predictor.  The goal here is to provide a typesafe way of getting an predictor
+     * which will return the correct output type given the command specified.
+     * <pre>
+     *  {@code
+     *      VWIntLearner vw = VWFactory.getVWLearner("--cb 4");
+     *  }
+     * </pre>
+     * @param command The VW initialization command.
+     * @param <T> The type of learner expected.  Note that this type implicitly specifies the output type of the learner.
+     * @throws ClassCastException If the specified type T is not a super type of the returned learner given the command.
+     * @return A VW Learner
+     */
+    @SuppressWarnings("unchecked")
+    public static <T extends VWLearner> T getVWLeaner(final String command) {
```

Comment:
This is super magic sauce that allows the return type to be specified at object creation time AND it'll fail if the supplied command does not return the correct type.